### PR TITLE
Basic support for embedded reference sequence in alignments

### LINKED ISSUE #1888: Implement default behaviour for alignments once reference_sequence is implemented
As discussed #146 we want to support the idea of reference sequences being associated with/stored in a tree sequence. The default behaviour of the alignments function should be updated to use this correctly.

A first pass at this might be (without knowing the full details of how we implement reference sequences yet)

- If no reference is associated with the tree sequence, fill in the gaps with the missing data character.
- Otherwise, use the reference sequence.

As of #1825 the default is to raise a ValueError if an explicit ``reference_sequence`` is not supplied.

### PR DESCRIPTION
**Pull request by jeromekelleher**
Closes #1888

Finally!

### CODE REVIEW

**Comment by jeromekelleher** [comment]
OK, this is ready for review I think!

I ended up doing a small bit of rearranging on the metadata docs as well, just to give me something to link to when discussing reference sequence metadata.

Pinging @hyanwong and @benjeffery for some review.

File: python/tskit/tables.py [hunk 1]
```diff
@@ -2684,13 +2710,22 @@ def time_units(self, time_units: str) -> None:
...
+        :ref:`reference sequence<sec_data_model_reference_sequence>`.
         """
         return bool(self._ll_tables.has_reference_sequence())
 
     @property
     def reference_sequence(self):
+        """
+        The :class:`.ReferenceSequence` associated with this :class:`.TableCollection`.
+
+        .. note:: Note that the behaviour of this attribute differs from
```
Comment (hyanwong) [inline python/tskit/tables.py:2723]: I don't really understand this, but 🤷

File: python/tskit/trees.py [hunk 2]
```diff
@@ -4886,7 +4893,7 @@ def genotype_matrix(
             isolated_as_missing=isolated_as_missing, alleles=alleles
         )
 
-    def alignments(self, *, reference_sequence=None):
+    def alignments(self, *, reference_sequence=None, missing_data_character=None):
         """
         Returns an iterator over the full sequence alignments for the samples in this
         tree sequence. Each alignment ``a`` is a string of length ``L`` where ``L``
```
Comment (hyanwong) [inline python/tskit/trees.py:4899]: Perhaps "``L`` is the sequence length of this tree sequence (which must have :ref:`discrete_genome` equal to ``True``), and ...

**Review by hyanwong — APPROVED** [review]
Looks OK to me, although I don't feel super qualified to check it all. Should we warn somewhere that we don't (yet) support alleles with len(allele) != 1?

File: python/tskit/trees.py (hunk 2 above)
Comment (jeromekelleher) [reply python/tskit/trees.py:4899]: Done.

**Comment by jeromekelleher** [comment]
updated, thanks @hyanwong

File: python/tskit/tables.py [hunk 3]
```diff
@@ -2500,11 +2500,31 @@ def __len__(self):
...
+
+    Metadata concerning reference sequences can be described using the
+    :attr:`.metadata_schema` and stored in the :attr:`.metadata` attribute.
+    See the :ref:`examples<sec_metadata_examples_reference_sequence>` for
+    idiomatic usage.
+
+    .. warning:: This API is preliminary and currently only supports accessing
+       reference sequence information via the ``.data`` attribute. Future versions
+       will also enable transparent fetching of known reference sequences
+       from a URL.
```
Comment (benjeffery) [inline python/tskit/tables.py:2515]: Would e nice to link to an issue here so we can gauge demand.

File: python/tskit/tables.py [hunk 4]
```diff
@@ -2500,11 +2500,31 @@ def __len__(self):
...
+
     def __init__(self, ll_reference_sequence):
         super().__init__(ll_reference_sequence)
         self._ll_reference_sequence = ll_reference_sequence
 
     def is_null(self) -> bool:
+        """
+        Returns True if this :class:`.ReferenceSequence` is null, i.e.,
+        corresponding to a :class:`.TableCollection` in which no reference sequence
+        has been specified.
```
Comment (benjeffery) [inline python/tskit/tables.py:2526]: Maybe be clearer here about what "specified" means? i.e. empty in all fields?

**Review by benjeffery — APPROVED** [review]
LGTM, a couple of doc comments

File: python/tskit/tables.py (hunk 3 above)
Comment (jeromekelleher) [reply python/tskit/tables.py:2515]: Done: #2022

**Comment by jeromekelleher** [comment]
OK, all done @benjeffery - I think we're ready to merge!

**Pull request merged**